app/controllers/admin/enterprises_controller.rb

```diff
@@ -113,6 +114,51 @@ def for_line_items
       end
     end
 
+    def stripe_connect
+      redirect_to authorize_stripe(params[:enterprise_id])
+    end
+
+    def stripe_connect_callback
+      if params["code"]
+        state = jwt_decode(params["state"])
+        unless state.keys.include? "enterprise_id"
+          redirect_to '/unauthorized' and return
+        end
+
+        # Get the Enterprise
+        @enterprise = Enterprise.find_by_permalink(state["enterprise_id"])
+        # Get the deets from Stripe
+        response_params = get_stripe_token(params["code"]).params
+
+        # In case of a problem, need to also issue a request to disconnect the account from Stripe
+        if !(spree_current_user.enterprises.include? @enterprise) && !(spree_current_user.admin?)
+          deauthorize_request_for_stripe_id(response_params["stripe_user_id"])
+          redirect_to '/unauthorized' and return
+        end
```

Comment:
I think that lines from 123 until here are business logic that would be better off in a factory method of `StripeAccount` at least, or even on a different object.

As I see it, we could set up error handling here in the controller to deal with these two exceptions:

```ruby
def stripe_connect_callback
  stripe_account = StripeAccountFactory.new(params)
  if stripe_account.save
    ...
  end
rescue Stripe::Error
  redirect_to '/unauthorized' and return
end
```

Reply:
Resolved